pyfr/solvers/euler/kernels/entropyfilter.mako

```diff
@@ -0,0 +1,158 @@
+# -*- coding: utf-8 -*-
+<%inherit file='base'/>
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+<%include file='pyfr.solvers.euler.kernels.entropy'/>
+
+<% inf = 1e20 %>
+<% ill_tol = 1e-6 %>
+<% zeta_tol = 1e-3 %>
+<% niters = 20 %>
+
+<%pyfr:macro name='get_minima' params='u, dmin, pmin, emin'>
+    fpdtype_t d, p, e;
+    fpdtype_t ui[${nvars}];
+
+    dmin = ${inf}, pmin = ${inf}, emin = ${inf};
+
+    for (int i = 0; i < ${nupts}; i++) {
+        % for j in range(nvars):
+        ui[${j}] = u[i][${j}];
+        % endfor
+
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p); emin = fmin(emin, e);
+    }
+
+    // If enforcing constraints on fpts/qpts, compute minima on fpts/qpts
+    % if con_fpts:
+    for (int i = 0; i < ${nfpts}; i++) {
+        for (int j = 0; j < ${nvars}; j++) {
+            ui[j] = ${pyfr.dot('intfpts[i][{k}]*u[{k}][j]', k=nupts)};
+        }
+
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        // Enforce only positivity constraints
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p);
+    }
+    % endif
+
+    % if con_qpts:
+    for (int i = 0; i < ${nqpts}; i++) {
+        for (int j = 0; j < ${nvars}; j++) {
+            ui[j] = ${pyfr.dot('intqpts[i][{k}]*u[{k}][j]', k=nupts)};
+        }
+
+        ${pyfr.expand('compute_entropy', 'ui', 'd', 'p', 'e')};
+        // Enforce only positivity constraints
+        dmin = fmin(dmin, d); pmin = fmin(pmin, p);
+    }
+    % endif
+</%pyfr:macro>
+
+<%pyfr:macro name='apply_filter' params='umodes, ffac, uf, zeta'>
+    // Precompute filter factors
+    % for i in range(nupts):
+    ffac[${i}] = exp(-zeta*${ubdegs2[i]});
+    % endfor
+
+    // Compute filtered solution
+    for (int uidx = 0; uidx < ${nupts}; uidx++) {
+        % for j in range(nvars):
+        uf[uidx][${j}] = 0.0;
+        % endfor
+
+        for (int midx = 0; midx < ${nupts}; midx++) {
+            for (int vidx = 0; vidx < ${nvars}; vidx++) {
+                tmp = ffac[midx]*umodes[midx][vidx]; // Filtered mode
```

Comment:
I think I would still need ffac (or something like it) to interface between the pyfr loop for ubdegs and the C loop for ffac.

Will look into the exp(-zeta) option

Reply:
Yes, you'll need to unroll the inner most loop (so it looks like a regular dot product).  This is probably worthwhile anyway.  Eliminating face will save you a nice amount in registers/local memory, and save flops (since everything will be factored).